sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RequestIdPolicy.java

```diff
@@ -3,24 +3,56 @@
 
 package com.azure.core.http.policy;
 
+import com.azure.core.http.HttpHeaders;
 import com.azure.core.http.HttpPipelineCallContext;
 import com.azure.core.http.HttpPipelineNextPolicy;
 import com.azure.core.http.HttpResponse;
 import reactor.core.publisher.Mono;
 
+import java.util.Objects;
 import java.util.UUID;
 
 /**
- * The pipeline policy that puts a UUID in the request header. Azure uses the request id as
+ * The pipeline policy that puts a UUID or user provided id in the request header. Azure uses the request id as
  * the unique identifier for the request.
+ * @see RequestIdPolicyOptions
  */
 public class RequestIdPolicy implements HttpPipelinePolicy {
     private static final String REQUEST_ID_HEADER = "x-ms-client-request-id";
 
+    private final RequestIdPolicyOptions requestIdPolicyOptions;
+
+    /**
+     * Creates {@link RequestIdPolicy} with default {@link ExponentialBackoff} as {@link RetryStrategy}and use
+     * 'retry-after-ms' in {@link HttpResponse} header for calculating retry delay.
+     */
+    public RequestIdPolicy() {
+        this.requestIdPolicyOptions = null;
+    }
+
+    /**
+     * Creates a {@link RequestIdPolicy} with the provided {@link RequestIdPolicyOptions}.
+     * User can specify {@link java.util.function.Function} to dynamically generated id for various id headers.
+     *
+     * @param requestIdPolicyOptions with given {@link RetryPolicyOptions}.
+     * @throws NullPointerException if {@code RequestIdPolicyOptions} is {@code null}.
+     */
+    public RequestIdPolicy(RequestIdPolicyOptions requestIdPolicyOptions) {
+        this.requestIdPolicyOptions = Objects.requireNonNull(requestIdPolicyOptions,
+            "'retryPolicyOptions' cannot be null.");
+    }
+
+
     @Override
     public Mono<HttpResponse> process(HttpPipelineCallContext context, HttpPipelineNextPolicy next) {
-        String requestId = context.getHttpRequest().getHeaders().getValue(REQUEST_ID_HEADER);
-        if (requestId == null) {
+        if (!Objects.isNull(requestIdPolicyOptions)) {
+            HttpHeaders httpHeaders =  requestIdPolicyOptions.getIdHeaderSupplier().get();
+            if (!Objects.isNull(httpHeaders)) {
+                httpHeaders.stream().forEach(httpHeader ->
+                    context.getHttpRequest().getHeaders().put(httpHeader.getName(), httpHeader.getValue())
+                );
+            }
+        } else {
             context.getHttpRequest().getHeaders().put(REQUEST_ID_HEADER, UUID.randomUUID().toString());
```

Comment:
This will now overwrite the requestId that was already in `context`. Existing requestId should be preserved. Random id should be used only when the requestId in context is null.